### PR TITLE
fix: sync ppm optimization

### LINKED ISSUE #10307: Sync++- Performance Optimization
The sync++- filter function is turning out to be very slow. We need a faster implementation in order to make sync++- practical.

### PR DESCRIPTION
**Pull request by cody-littley**
Closes #10307

### CODE REVIEW

File: platform-sdk/swirlds-platform-core/src/main/java/com/swirlds/platform/components/EventIntake.java [hunk 1]
```diff
@@ -190,6 +190,8 @@ public void addEvent(final EventImpl event) {
             phaseTimer.activatePhase(EventIntakePhase.ADDING_TO_HASHGRAPH);
             final List<ConsensusRound> consRounds = consensus().addEvent(event);
 
+            shadowGraph.setMinimumGenerationNonAncient(consensus().getMinGenerationNonAncient());
```
Comment (lpetrovic05) [inline platform-sdk/swirlds-platform-core/src/main/java/com/swirlds/platform/components/EventIntake.java:193]: does the same thing need to happen with the new intake?

File: platform-sdk/swirlds-platform-core/src/main/java/com/swirlds/platform/gossip/shadowgraph/ShadowGraph.java [hunk 2]
```diff
@@ -108,39 +116,65 @@ public class ShadowGraph implements Clearable {
...
      */
-    public ShadowGraph(final SyncMetrics syncMetrics) {
-        this(syncMetrics, -1);
-    }
+    private final TipsetTracker tipsetTracker;
+
+    /**
+     * The tipset of the latest self event, or null if there have been no self events.
+     */
+    private Tipset latestSelfEventTipset;
```
Comment (lpetrovic05) [inline platform-sdk/swirlds-platform-core/src/main/java/com/swirlds/platform/gossip/shadowgraph/ShadowGraph.java:129]: This complicates the shadowgraph. Should we maybe go with a wrapper or extension that has a tipset?

File: platform-sdk/swirlds-platform-core/src/main/java/com/swirlds/platform/gossip/shadowgraph/SyncUtils.java [hunk 3]
```diff
@@ -446,12 +432,17 @@ public static List<EventImpl> filterLikelyDuplicates(
                 continue;
             }
 
+            // TODO make sure this doesn't break when AB size changes
```
Comment (alittley) [inline platform-sdk/swirlds-platform-core/src/main/java/com/swirlds/platform/gossip/shadowgraph/SyncUtils.java:435]: TODO

**Review by alittley — APPROVED** [review]

File: platform-sdk/swirlds-platform-core/src/main/java/com/swirlds/platform/components/EventIntake.java (hunk 1 above)
Comment (cody-littley) [reply platform-sdk/swirlds-platform-core/src/main/java/com/swirlds/platform/components/EventIntake.java:193]: Ah, yes. You are correct. Having two pathways makes things so much more confusing. We would have noticed this pretty quickly in a longer test, symptom would have been an OOM. Apparently wasn't a fast enough leak to sink our 5 minute abbreviated panel tests.

File: platform-sdk/swirlds-unit-tests/core/swirlds-platform-test/src/test/java/com/swirlds/platform/test/sync/SyncFilteringTest.java [hunk 4]
```diff
@@ -141,8 +141,8 @@ void filterLikelyDuplicatesTest() {
                 }
             }
 
-            final List<EventImpl> filteredEvents =
-                    SyncUtils.filterLikelyDuplicates(shadowGraph, selfId, nonAncestorThreshold, clock.now(), allEvents);
+            final List<EventImpl> filteredEvents = SyncUtils.filterLikelyDuplicates(
+                    selfId, nonAncestorThreshold, clock.now(), allEvents, null); // TODO
```
[the remark this answers is not in the document]
Comment (cody-littley) [reply platform-sdk/swirlds-unit-tests/core/swirlds-platform-test/src/test/java/com/swirlds/platform/test/sync/SyncFilteringTest.java:145]: Removed TODO. Wasn't actually causing the test to fail, I just needed to figure out if it was problematic.

**Review by alittley — DISMISSED** [review]

File: platform-sdk/swirlds-platform-core/src/main/java/com/swirlds/platform/gossip/shadowgraph/ShadowGraph.java (hunk 2 above)
Comment (cody-littley) [reply platform-sdk/swirlds-platform-core/src/main/java/com/swirlds/platform/gossip/shadowgraph/ShadowGraph.java]: Per our discussion, I've removed this from `ShadowGraph.java` and maintain the data structure in an adjacent location.

File: platform-sdk/swirlds-platform-core/src/main/java/com/swirlds/platform/gossip/shadowgraph/LatestEventTipsetTracker.java [hunk 5]
```diff
@@ -0,0 +1,94 @@
...
+import edu.umd.cs.findbugs.annotations.NonNull;
+import edu.umd.cs.findbugs.annotations.Nullable;
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Objects;
+
+/**
+ * Tracks the tipset of the latest self event. Must be thread safe, gossip will access it from multiple threads.
+ */
+public class LatestEventTipsetTracker { // TODO add a unit test
```
Comment (edward-swirldslabs) [inline platform-sdk/swirlds-platform-core/src/main/java/com/swirlds/platform/gossip/shadowgraph/LatestEventTipsetTracker.java]: TODO in code.

**Review by edward-swirldslabs — APPROVED** [review]

File: platform-sdk/swirlds-platform-core/src/main/java/com/swirlds/platform/gossip/shadowgraph/LatestEventTipsetTracker.java (hunk 5 above)
Comment (cody-littley) [reply platform-sdk/swirlds-platform-core/src/main/java/com/swirlds/platform/gossip/shadowgraph/LatestEventTipsetTracker.java]: Lol, I removed this two minutes after the original push. Thanks for the prompt review, I guess! 😺

File: platform-sdk/swirlds-platform-core/src/main/java/com/swirlds/platform/gossip/shadowgraph/LatestEventTipsetTracker.java [hunk 6]
```diff
@@ -0,0 +1,94 @@
...
+    private Tipset latestSelfEventTipset;
+
+    /**
+     * Constructor.
+     *
+     * @param time        provides wall clock time
+     * @param addressBook the current address book
+     */
+    public LatestEventTipsetTracker(
+            @NonNull final Time time, @NonNull final AddressBook addressBook, @NonNull final NodeId selfId) {
```
Comment (alittley) [inline platform-sdk/swirlds-platform-core/src/main/java/com/swirlds/platform/gossip/shadowgraph/LatestEventTipsetTracker.java:48]: missing param doc for `selfId`

**Review by alittley — DISMISSED** [review]

File: platform-sdk/swirlds-platform-core/src/main/java/com/swirlds/platform/gossip/shadowgraph/LatestEventTipsetTracker.java (hunk 6 above)
Comment (cody-littley) [reply platform-sdk/swirlds-platform-core/src/main/java/com/swirlds/platform/gossip/shadowgraph/LatestEventTipsetTracker.java:48]: added

**Review by alittley — APPROVED** [review]

**Pull request merged**